airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/partition.py

```diff
@@ -0,0 +1,30 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, Mapping
+
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+
+
+class Partition(ABC):
```

Comment:
[question from](https://github.com/airbytehq/airbyte/pull/30113/files#r1328245203) @sherifnada 

> I think if a Partition owns the read method then it will be hard to globally coordinate request concurrency across partitions e.g: what if a partition wants to dynamically resize its request to multiple requests? What's the rationale behind this having here, and is it possible you think to move it to the stream class?

Reply:
@sherifnada rephrasing the scenario to make sure I understand. Please correct me if I get it wrong.

1. We want to fetch a large partition (let's say 2021-01-01 to 2021-12-31)
2. The request fails because the query is too large (or we want to optimize it ahead of time)
3. We submit multiple small requests instead ([2021-01-01 to 2021-06-30] and [2021-07-01 to 2021-12-31])

Since we decided to parallelize at the partition level and not at the stream level, I think this should be implemented in the partition itself. 


```
class PartitionWithCustomChunking(Partition):

  def __init__(requester, parser, start_time, end_time):
    ...

  def read():
    request = self._generate_request(start_time, end_time)
    try:
      response = self._requester.submit(request)
    except:
      requests = self._split(request)
      for r in requests:
        response = self._requester.submit(r)
        yield from self._parser(response)
```

I'm open to renaming the class to make it clear this is not just a descriptor, but the reason I don't think this belongs in the `Stream` because the requesting is tangential to the parallelism.

Reply:
(@sherifnada feel free to chime in if you disagree)
@maxi297 my understanding is:

some facebook streams would have non-trivial partitions.
For instance, we sometimes [split the job into two in the hope the job failed because we were requesting too much data](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/async_job_manager.py#L107-L110).

It would make sense to say that since if the two jobs are independent (eg if they're non-overlapping time ranges), they could be processed concurrently.

I don't think this is a use case worth optimizing for for 2 reasons:
1. It's pretty niche, so the cost of processing the two smaller job sequentially doesn't seem too bad
2. we could implement a partition object that would be able to process the two smaller job sequentially by giving it access to a threadpool